<div style="border:solid Chocolate 2px; padding: 40px">
    
 <b>Добрый день, Софья!</b>
    
Меня зовут Ксения Цыганова, и я буду комментировать твою работу и помогать тебе сделать ее сильнее. В Практикуме мы общаемся на «ты», но если тебе это будет некомфортно, пожалуйста, скажи, и мы будем обращаться друг к другу на «вы».
    
Все мои комментарии делятся на три вида:

    
👍 Здесь все отлично, идем дальше! - <font color='green'> зеленый цвет </font>
    
💡 Все хорошо, однако, можно оптимизировать или рассмотреть другие способы (необязательно для исполнения, чтобы пройти дальше) - <font color='gold'>желтый цвет</font>
    
🤔 Необходимо внести исправления, без которых я не смогу принять проект - <font color='red'>красный цвет</font>
    
    
Свои комментарии я буду выделять заголовком «комментарий ревьюера». Пожалуйста, не удаляй их, это поможет сократить время проверки на повторных ревью.Ты также можешь писать свои комментарии к моим и выделять их любым понравившимся способом, отличным от моего. Если у тебя будут возникать вопросы в течение нашей работы, я с радостью на них отвечу.

Цель моего ревью не критика и не возможность дать тебе готовые ответы, это способ привести тебя в нужное русло, чтобы ты смогла развить собственные идеи и решения. Твой будущий работодатель будет поступать похожим образом. 
        
   <b>У тебя все получится!</b>

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 
    
Отлично, что есть вступительный комментарий и содержание.

## Загрузка данных

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.pipeline import Pipeline

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

    
<b>Все отлично!👍:</b> 
    
Отлично, что все импорты собраны в начале проекта. Это поможет в случае чего сразу увидеть, какие библиотеки следует доустановить.

In [2]:
data = pd.read_csv('/datasets/insurance.csv')
print(data.head())
data.info()

   Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 
        
Для загрузки данных можно использовать библиотеку os или констукцию try/except. Это поможет работать с проектом как локально, так и через Jupyter Hub, не внося дополнительных правок.

In [3]:
data.describe()

Пол      Возраст      Зарплата  Члены семьи  Страховые выплаты
count  5000.000000  5000.000000   5000.000000  5000.000000        5000.000000
mean      0.499000    30.952800  39916.360000     1.194200           0.148000
std       0.500049     8.440807   9900.083569     1.091387           0.463183
min       0.000000    18.000000   5300.000000     0.000000           0.000000
25%       0.000000    24.000000  33300.000000     0.000000           0.000000
50%       0.000000    30.000000  40200.000000     1.000000           0.000000
75%       1.000000    37.000000  46600.000000     2.000000           0.000000
max       1.000000    65.000000  79000.000000     6.000000           5.000000

In [4]:
data[['Возраст']] = data[['Возраст']].astype(int)

**Код не содержит пропуски, изменили тип столбца "возраст" на INT**

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 

Молодец, что изучаешь данные, с которыми работаешь.

## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y) 
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y 
$$

**Вопрос:** Изменится ли качество линейной регрессии при умножении признаков на обратимую матрицу?

**Ответ:** Не изменится

**Обоснование:** 

Используя свойства матриц 
Домножим сначала матрицу Х на единичную,которую раскроем как произведение матрицы Р на обратную

$$
a = Xw = XEw = XPP^{-1}w= XPw'
$$

Получилось,что $a=XPw'$ и $a=Xw$ => $w'=Р^{-1}w$,для подтверждения тождества подставим в формулу обучения и получим:

$$
w' = ((XP)^TXP)^{-1}(XP)^Ty = (P^TX^TXP)^{-1}P^TX^Ty =
P^{-1}(X^TX)^{-1}(P^T)^{-1}P^TX^Ty= P^{-1}(X^TX)^{-1}EX^Ty =
P^{-1}(X^TX)^{-1}X^Ty =  P^{-1}w
$$

Что и требовалось доказать $XPw'=Xw$.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 
    
Теоретическое обоснование и матричные преобразования корректны, молодец!

## Алгоритм преобразования

**Алгоритм**

1. Разбиваем датасет на выборки.
2. Создаем квадратную обратимую матрицу
3. Перемножаем матрицы, тем самым получаем преобразованные признаки
4. Обучаем модели
5. Ищем метрику качества на исходных и  преобразованных данных
6. Выводим результаты для сравнения и подведения итогов

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 
    
Отлично, что есть описание алгоритма.

**Обоснование**
Умножать признаки будем на одну и ту же матрицу, соблюдая условие для умножения матриц(совпадения кол-во столбцов), в результате будет произведение таблицы с признаками на вектор. Проверим это в 4 пункте.

## Проверка алгоритма


<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
    
Здесь код выдает ошибку. Исправь, пожалуйста, продолжу ревью на следующей итерации.

In [5]:
features = data.drop('Страховые выплаты',axis=1)
target = data['Страховые выплаты'] #целевой признак

In [6]:
features_train, features_test, target_train, target_test = train_test_split(
    features, target, test_size=0.25, random_state=12345)

<div class="alert alert-success">
<h2> Комментарий ревьюера №2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 

Молодец, что разбиваешь данные на выборки

Посчитаем метрику R2 для отмасштабированной модели

In [7]:
model = LinearRegression()
scaller = StandardScaler()
pipeline = Pipeline([("standard_scaller", scaller),("linear_regression", model)])
pipeline.fit(features_train, target_train)
R2_data = r2_score(target_test, pipeline.predict(features_test))
print("R2 =", R2_data)

R2 = 0.4352275712702668


<div class="alert alert-success">
<h2> Комментарий ревьюера №2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 

Молодец, что используешь pipeline в сочетании со StandardScaler

**Метрика R2 модели без преобразования равна 0.4352275712702668**

In [8]:
# Создадим матрицу
count = features.shape[1]
matrix = np.random.randint(1,10, (count,count))
display(features.head())
m_features = features @ matrix
np.linalg.inv(matrix)
display(m_features.head())
matrix

Пол  Возраст  Зарплата  Члены семьи
0    1       41   49600.0            1
1    0       46   38000.0            1
2    0       29   21000.0            0
3    0       21   41700.0            2
4    1       28   26100.0            0

0         1         2         3
0  446703.0  198572.0  297736.0  297687.0
1  342329.0  152188.0  228144.0  228095.0
2  189203.0   84116.0  126087.0  126058.0
3  375461.0  166892.0  250275.0  250248.0
4  235105.0  104516.0  156691.0  156658.0

array([[9, 4, 7, 2],
       [7, 4, 3, 2],
       [9, 4, 6, 6],
       [7, 4, 6, 3]])


<div class="alert alert-success">
    
<h2> Комментарий ревьюера №2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 
    
Отлично, что есть проверка случайной матрицы на обратимость.

Произведем рассчет метрики модели с преобразованием

In [9]:
features_train, features_test, target_train, target_test = train_test_split(
    m_features, target, test_size=0.25, random_state=12345)

In [10]:
model = LinearRegression()
scaller = StandardScaler()
pipeline = Pipeline([("standard_scaller", scaller),("linear_regression", model)])
pipeline.fit(features_train, target_train)
R2_m_data = r2_score(target_test, pipeline.predict(features_test))
print("R2 =", R2_m_data)

R2 = 0.43522757127006306


**Метрика R2 модели с преобразованием равна 0.4352275712701458**

<div class="alert alert-success">
    
<h2> Комментарий ревьюера №2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 
    
Молодец, что сравниваешь качество работы линейной модели до и после шифрования обучающих данных.

In [11]:
display(m_features.head()) #данные после перобразования
features.head() #исходнные данные 

0         1         2         3
0  446703.0  198572.0  297736.0  297687.0
1  342329.0  152188.0  228144.0  228095.0
2  189203.0   84116.0  126087.0  126058.0
3  375461.0  166892.0  250275.0  250248.0
4  235105.0  104516.0  156691.0  156658.0

Пол  Возраст  Зарплата  Члены семьи
0    1       41   49600.0            1
1    0       46   38000.0            1
2    0       29   21000.0            0
3    0       21   41700.0            2
4    1       28   26100.0            0

In [12]:
matrix = np.linalg.inv(matrix)
recov_features = m_features @ matrix
recov_features = round(recov_features,2)
recov_features.head() #исходные данные

0     1        2    3
0  1.0  41.0  49600.0  1.0
1 -0.0  46.0  38000.0  1.0
2 -0.0  29.0  21000.0  0.0
3 -0.0  21.0  41700.0  2.0
4  1.0  28.0  26100.0  0.0

<div class="alert alert-warning"> 
    Используя исходные данные, с помощью преобразований получили первоначальные значения. 

<div class="alert alert-success">
<h2> Комментарий ревьюера №3<a class="tocSkip"> </h2>

<b>Все отлично!👍</b> 

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера №2<a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
    
Также стоит проверить возможность восстановления первоначальных данных после шифрования.

### Вывод

- В данной работе мы загрузили данные, качество данных адекватное, для экономии памяти мы заменили тип столбца "Возраст" на целочисленный
- В результате был разработан алгоритм, где данные были преобразованы методом умножения признаков исходных данных на обратимую матрицу
- Качество линейной регресии не изменилось от использования исxодной матрицы и исходной матрицы, умноженную на обратимую 
- Произвели доказательство нашей теории, основываясь на законах ленейной алегебры, а так же с помощью кода доказали это с использованием метрики R2.
-  Метрика R2 модели без преобразования и метрика R2 модели с преобразованием совпали, (учитывая, что значения 12-ой цифры после запятой не считали)

<div class="alert alert-warning">
    <h2> Комментарий ревьюера №2<a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Здесь я оставлю некоторые источники, которые могут быть интересны и полезны:

1. Теория по обратным матрицам: https://www.berdov.com/works/matrix/obratnaya-matrica/ 
    
Если интересна тема шифрования, то можно почитать:

2. Обзоры по применению криптографии в машинном обучении: 
    https://arxiv.org/pdf/1902.04109.pdf, 
    https://www.researchgate.net/publication/342746786_Some_Applications_of_Machine_Learning_in_Cryptography
        
3. Подробнее про аналогичный алгоритм: https://www.cs.cmu.edu/~rjhall/JOS_revised_May_31a.pdf

<div style="border:solid Chocolate 2px; padding: 40px">
    
<b>Общий вывод по проекту</b>

Софья, спасибо за работу!

**Положительные моменты проекта, которые хочу еще раз подчеркнуть**:

1. Данные загружены и изучены
    
2. Линейная модель обучена верно
    
3. Предложен алгоритм кодирования
    
4. Проверен тот факт, что шифрование не влияет на качество линейной регрессии
    
5. Есть проверка случайной матрицы на обратимость
    
6. Теоретическое обоснование алгоритма шифрования корректное
    
      

**Замечания, на которые стоит обратить внимания при будущей работе**:
    

    1. Проверить возможность восстановления данных после применения алгоритма


    
Остальные мои комментарии ты найдешь выше.
    
**Желаю удачи и жду твой проект на повторное ревью!**

<div style="border:solid Chocolate 2px; padding: 40px">
    
<b>Общий вывод по проекту №2</b>

Софья, спасибо за работу,
    
**все хорошо, проект принимаю и желаю успехов в дальнейшем!**

## Чек-лист проверки

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [ ]  Весь код выполняется без ошибок
- [ ]  Ячейки с кодом расположены в порядке исполнения
- [ ]  Выполнен шаг 1: данные загружены
- [ ]  Выполнен шаг 2: получен ответ на вопрос об умножении матриц
    - [ ]  Указан правильный вариант ответа
    - [ ]  Вариант обоснован
- [ ]  Выполнен шаг 3: предложен алгоритм преобразования
    - [ ]  Алгоритм описан
    - [ ]  Алгоритм обоснован
- [ ]  Выполнен шаг 4: алгоритм проверен
    - [ ]  Алгоритм реализован
    - [ ]  Проведено сравнение качества моделей до и после преобразования